### ISSUE TITLE
Failed: Detected file entries with not associated blocks

### ISSUE THREAD

**Issue by kaasszje** [duplicati/duplicati#1695]
I've updated to the new version yesterday: 2.0.1.7_experimental_2016-03-19

Now after my daily backup it stops with this error.

```
Failed: Detected file entries with not associated blocks
Details: System.Exception: Detected file entries with not associated blocks
  at Duplicati.Library.Main.Database.LocalDatabase.VerifyConsistency (IDbTransaction transaction, Int64 blocksize, Int64 hashsize) [0x00000] in <filename unknown>:0
  at Duplicati.Library.Main.Operation.BackupHandler.Run (System.String[] sources, IFilter filter) [0x00000] in <filename unknown>:0
  at Duplicati.Library.Main.Controller+<Backup>c__AnonStorey0.<>m__0 (Duplicati.Library.Main.BackupResults result) [0x00000] in <filename unknown>:0
  at Duplicati.Library.Main.Controller.RunAction[BackupResults] (Duplicati.Library.Main.BackupResults result, System.String[]& paths, IFilter& filter, System.Action`1 method) [0x00000] in <filename unknown>:0
```
##

**Comment by kaasszje** [comment]
Any idea what I can do to resolve this ?

**Comment by kt368** [comment]
I have same error. Version 2.0.1.11- 2.0.1.11_experimental_2016-04-08.

**Comment by kaasszje** [comment]
I've tried to repair the databases.
On one of the backup sets it worked.
The other however doesn't finish the database restore.
This now gives the error: 
System.Exception: The database was attempted repaired, but the repair did not complete. This database may be incomplete and the backup process cannot continue. You may delete the local database and attempt to repair it again.
  at Duplicati.Library.Main.Operation.BackupHandler.Run (System.String[] sources, IFilter filter) [0x00000] in <filename unknown>:0 
  at Duplicati.Library.Main.Controller+<Backup>c__AnonStorey0.<>m__0 (Duplicati.Library.Main.BackupResults result) [0x00000] in <filename unknown>:0 
  at Duplicati.Library.Main.Controller.RunAction[BackupResults] (Duplicati.Library.Main.BackupResults result, System.String[]& paths, IFilter& filter, System.Action`1 method) [0x00000] in <filename unknown>:0

**Comment by kenkendk** [comment]
The error means that the database has an error. Previously there was no check for this, but it means that some files in your database cannot be restored. It also means that the remote data may have recorded these files incorrectly, so they cannot be restored.

If you can create a bugreport, I can examine the database and find out what files are involved.

If you prefer to just move on, you can delete (or rename) the local database, and then run "repair" to get it recreated. 

If your remote data is broken, the restored database will also be in a broken state. If this happens, I would like a bugreport from the recreated database to see if I can figure out a workaround to the problem.

I am not sure what causes the SFTP error.

**Comment by kaasszje** [comment]
Checked the file system log of duplicati: 
2016-04-17 13:17:14Z - Error: Remote file referenced as duplicati-be3c96445629c469f85249fbbc2b1f755.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:17:17Z - Error: Remote file referenced as duplicati-bda7117e5d15245428ec82e02396980d0.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:18:14Z - Error: Remote file referenced as duplicati-b3facaff47a5c42bd89f5881ad4c7a1dd.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:19:44Z - Error: Remote file referenced as duplicati-b7ecdcc9b92614c4985f2bc0912484711.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:20:25Z - Error: Remote file referenced as duplicati-b8aa625b061ef4a0183a3f5c1a7848538.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:21:36Z - Error: Remote file referenced as duplicati-b858fc42358b24157bd420ccf490b1d93.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:22:21Z - Error: Remote file referenced as duplicati-bb3123474b30f459dbdf0943e748ca476.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:24:35Z - Error: Remote file referenced as duplicati-b27d07577203944dca29f52a903d09639.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:26:05Z - Error: Remote file referenced as duplicati-b8cdee0c2a737477ba35d9c8fc4a391ae.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:41:06Z - Error: Remote file referenced as duplicati-b7eda40e3ec614ec792124875d7e8d842.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 13:47:04Z - Error: Remote file referenced as duplicati-b6ba38100cbf14b038514c41d9bee4291.dblock.zip.aes, but not found in list, registering a missing remote file
2016-04-17 23:25:44Z - Error: Terminated with error: No such file
Renci.SshNet.Common.SftpPathNotFoundException: No such file
  at Renci.SshNet.Sftp.SftpSession.RequestOpen (System.String path, Flags flags, Boolean nullOnError) [0x00000] in <filename unknown>:0
  at Renci.SshNet.SftpClient.InternalDownloadFile (System.String path, System.IO.Stream output, Renci.SshNet.Sftp.SftpDownloadAsyncResult asyncResult, System.Action`1 downloadCallback) [0x00000] in <filename unknown>:0
  at Renci.SshNet.SftpClient.DownloadFile (System.String path, System.IO.Stream output, System.Action`1 downloadCallback) [0x00000] in <filename unknown>:0
  at Duplicati.Library.Backend.SSHv2.Get (System.String remotename, System.IO.Stream stream) [0x00000] in <filename unknown>:0
  at Duplicati.Library.Main.BackendManager.coreDoGetPiping (Duplicati.Library.Main.FileEntryItem item, IEncryption useDecrypter, System.Int64& retDownloadSize, System.String& retHashcode) [0x00000] in <filename unknown>:0

**Comment by kaasszje** [comment]
For the brokendb bugreport: https://fam-krijgsman.nl/sharing/KzpA20fPq
(Sorry for the link, but it's arround 200mb and github wouldn't accept that ;))

I'll see if I still have the old db, and upload a report later on as well.

**Comment by ulicky** [comment]
Hi I have same error in #1678, I think, it has similar root cause. These errors needs some repair not on local database but on remote side.

**Comment by ulicky** [comment]
Hi, any progress on that? It prevent from upgrades from 2.0.1.9 :(

**Comment by 4m1g0** [comment]
This problem can be reproduced with this steps:
- Use ngax theme
- Backup some folder
- Use the main menu option to restore the backup (don't use the backup entry from the list) instead use the menu option (screenshot attached)
- An error pops up every time saying: Detected file entries with not associated blocks

The backup is restored correctly anyway.

This was caused by this commit https://github.com/duplicati/duplicati/commit/b92f52331ea7112d099d0d913a9d72e1986793a0 and the problem persists until master.

I think the problem is that when restoring a backup with a temporary backup, this condition is never satisfied because we don't have downloaded all the files from the storage, so we don't have the block in the local storage.

I don't understand this process perfectly, but I think this check should be done only for local backups (not temporary) or maybe we should do it later.

![image](https://cloud.githubusercontent.com/assets/204293/18445753/60409a80-791f-11e6-95be-f2330851e2ea.png)

![image](https://cloud.githubusercontent.com/assets/204293/18446109/12804eb0-7921-11e6-948a-454c218fa5c9.png)

**Comment by mnaiman** [comment]
I think commit is good even it broke my backup too. 
I had to repair remote backup data and repair database.
Now it is working.
For some reason some of the 2015 backups were affected by this, but it is repairable.

**Comment by 4m1g0** [comment]
Just to make it more clear.

This call here: https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Main/Operation/RecreateDatabaseHandler.cs#L430

We call verifyConsistency which checks for blocks in the database, but we know there are no blocks at all because at this point we only have restored the dlist volume in order to list the files and versions before restore.

I am not sure how we should handle this. 
Am I on the right path?

**Comment by kenkendk** [comment]
@4m1g0 : I think you are correct, a partially re-created database is (obviously) not complete, and hence fails the check.

Maybe we need to pass another flag to indicate that the database is partially recreated, as is done with marking the database incomplete.

**Comment by kenkendk** [comment]
After re-reading this, there are multiple issues mixed in this thread:

@4m1g0 : You are correct, and the bug is reproducible. I have made a fix that solves this problem in 490776ed73834e788ec56644c0cb4e2f537e3f87.

@mnaiman : Your issue is related to a missing remote file. Recovering from this situation is not yet supported, as it involves loosing some information, or patching files with zero blocks. I think this is with #1503.

@kaasszje : Unfortunately, this does not solve your issue, but I have now downloaded the database and am looking through it.
